### PR TITLE
[DNM] server: send snapshot via file

### PR DESCRIPTION
**Pull request by BusyJay**
This is a partial implement for #594.

@ngaut @siddontang @disksing @tiancaiamao PTAL

### CODE REVIEW

File: src/server/server.rs [hunk 1]
```diff
@@ -549,12 +528,16 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver> Handler for Server<T, S> {
         if events.is_writable() {
             self.on_writable(event_loop, token);
         }
+
+        if events.is_hup() {
```
Comment (siddontang) [inline src/server/server.rs]: if events still has readable, we can't remove conn here.

File: src/raftstore/store/config.rs [hunk 2]
```diff
@@ -64,6 +64,9 @@ pub struct Config {
     pub region_check_size_diff: u64,
     pub pd_heartbeat_tick_interval: u64,
     pub pd_store_heartbeat_tick_interval: u64,
+
+    /// Directory for snapshot files
+    pub snap_path: String,
```
Comment (ngaut) [inline src/raftstore/store/config.rs]: s/snap_path/snap_dir/

File: src/server/snap.rs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    if !snap_file.exists() {
+        return Err(box_err!("missing snap file: {:?}", snap_file.path()));
+    }
+    // snapshot file has been validated when created, so no need validate again.
+
+    let mut f = try!(File::open(snap_file.path()));
+    snap_file.delete_when_drop();
+    let mut conn = try!(TcpStream::connect(&addr));
+
+    rpc::encode_msg(&mut conn, data.msg_id, &data.msg)
```
Comment (siddontang) [inline src/server/snap.rs]: we can log the time for sending snapshot file with size.

File: src/server/snap.rs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    let snap = data.msg.get_raft().get_message().get_snapshot();
+    let mut snap_file = box_try!(SnapFile::from_snap(snap_dir.as_path(), &snap, SNAP_GEN_PREFIX));
+    if !snap_file.exists() {
+        return Err(box_err!("missing snap file: {:?}", snap_file.path()));
+    }
+    // snapshot file has been validated when created, so no need validate again.
+
+    let mut f = try!(File::open(snap_file.path()));
+    snap_file.delete_when_drop();
```
Comment (siddontang) [inline src/server/snap.rs]: if send failed, the snapshot will be deleted?

I guess we may still use this term + index for snapshot.

Comment (BusyJay) [inline src/server/snap.rs]: This may lead to snapshot file leak.

Comment (siddontang) [inline src/server/snap.rs]: We can introduce GC to fix this leak and we should also use it.
E.g, we panic when we sending the snapshot, so how to delete this snapshot file?

Comment (BusyJay) [inline src/server/snap.rs]: Cleanup will be done when restart.

File: src/raftstore/store/config.rs [hunk 5]
```diff
@@ -98,6 +102,9 @@ impl Config {
             return Err(box_err!("raft log gc threshold must >= 1, not {}",
                                 self.raft_log_gc_threshold));
         }
+        if self.snap_path == "" {
```
Comment (ngaut) [inline src/raftstore/store/config.rs]: is_empty()

File: src/bin/tikv-server.rs [hunk 6]
```diff
@@ -318,14 +320,37 @@ fn get_store_path(matches: &Matches, config: &toml::Value) -> String {
...
 
+fn get_snap_path(matches: &Matches, config: &toml::Value) -> String {
+    let path = get_string_value("",
+                                "server.snapshot",
+                                matches,
+                                config,
+                                Some(TEMP_DIR.to_owned()),
+                                |v| v.as_str().map(|s| s.to_owned()));
+    if path == TEMP_DIR {
+        return path;
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: TMEP_DIR is "", maybe we should return "/tmp" instead.

File: src/raftstore/store/peer_storage.rs [hunk 7]
```diff
@@ -503,9 +531,213 @@ impl PeerStorage {
...
+    /// Validate whether current file is broken.
+    pub fn validate(&self) -> io::Result<()> {
+        let mut reader = try!(File::open(self.path()));
+        let mut digest = Digest::new(crc32::IEEE);
+        let len = try!(reader.metadata()).len();
+        if len < 4 {
+            return Err(io::Error::new(ErrorKind::InvalidInput, format!("file length {} < 4", len)));
+        }
+        let to_read = len as usize - 4;
+        let mut total_readed = 0;
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: total_read

File: src/raftstore/store/peer_storage.rs [hunk 8]
```diff
@@ -503,9 +531,213 @@ impl PeerStorage {
...
+        let mut buffer = vec![0; 1024];
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: 1k seems a little small here. 
4k or 8k may be better.

File: src/server/snap.rs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+                        }
+                    }
+                    None => error!("invalid snap token {:?}", token),
+                }
+            }
+            Task::Close(token) => {
+                match self.files.remove(&token) {
+                    Some((mut writer, msg)) => {
+                        if let Err(e) = writer.save() {
+                            error!("failed to save file {:?}: {:?}", token, e);
```
Comment (siddontang) [inline src/server/snap.rs]: if save failed, no need return?

File: src/server/snap.rs [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+const DEFAULT_SENDER_POOL_SIZE: usize = 3;
+
+/// `Task` that `Runner` can handle.
+///
+/// `Register` register a pending snapshot file with token;
+/// `Write` write data to snapshot file;
+/// `Close` save the snapshot file;
+/// `Discard` discard all the unsaved changes made to snapshot file;
+/// `SendTo` send the snapshot file to specified address.
+pub enum Task {
```
Comment (ngaut) [inline src/server/snap.rs]: I suggest rename Task to BgSnapTask. Any ideas ?

File: src/server/conn.rs [hunk 11]
```diff
@@ -114,48 +130,145 @@ impl Conn {
...
+        }
 
-            // payload here can't be None.
-            let mut payload = self.payload.take().unwrap();
-            try!(try_read_data(&mut self.sock, &mut payload));
-            if payload.remaining() > 0 {
-                // we need to read more data for payload
-                self.payload = Some(payload);
-                break;
+        payload = if self.read_size == self.file_size {
```
Comment (siddontang) [inline src/server/conn.rs]: where to update read_size?

File: src/server/conn.rs [hunk 12]
```diff
@@ -114,48 +130,145 @@ impl Conn {
...
+            create_mem_buf(1)
```
Comment (siddontang) [inline src/server/conn.rs]: why create extra 1 here?

File: src/server/server.rs (hunk 1 above)
Comment (BusyJay) [inline src/server/server.rs]: If it's readable, all the data has been read via `self.on_readable` before reaching this line.

File: src/server/conn.rs (hunk 12 above)
Comment (BusyJay) [inline src/server/conn.rs]: To break the while loop outside. Never mind, I inline this function and just break now.

File: src/server/snap.rs [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    if !snap_file.exists() {
+        return Err(box_err!("missing snap file: {:?}", snap_file.path()));
+    }
+    // snapshot file has been validated when created, so no need validate again.
+
+    let mut f = try!(File::open(snap_file.path()));
+    snap_file.delete_when_drop();
+    let mut conn = try!(TcpStream::connect(&addr));
+
+    let res = rpc::encode_msg(&mut conn, data.msg_id, &data.msg)
```
Comment (tiancaiamao) [inline src/server/snap.rs]: using try! macro here maybe clearer?

File: src/server/snap.rs [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+    let snap = data.msg.get_raft().get_message().get_snapshot();
+    let mut snap_file = box_try!(SnapFile::from_snap(snap_dir.as_path(), &snap, SNAP_GEN_PREFIX));
+    if !snap_file.exists() {
+        return Err(box_err!("missing snap file: {:?}", snap_file.path()));
+    }
+    // snapshot file has been validated when created, so no need validate again.
+
+    let mut f = try!(File::open(snap_file.path()));
+    snap_file.delete_when_drop();
+    let mut conn = try!(TcpStream::connect(&addr));
```
Comment (siddontang) [inline src/server/snap.rs]: set no delay for conn.

**Comment by BusyJay** [comment]
PTAL

File: src/server/conn.rs [hunk 15]
```diff
@@ -114,48 +130,141 @@ impl Conn {
         Ok(())
     }
 
-    pub fn read<T, S>(&mut self, _: &mut EventLoop<Server<T, S>>) -> Result<Vec<ConnData>>
+
+    pub fn on_readable<T, S>(&mut self,
```
Comment (siddontang) [inline src/server/conn.rs]: please change write -> on_writable too.

File: src/raftstore/store/peer_storage.rs [hunk 16]
```diff
@@ -503,9 +531,213 @@ impl PeerStorage {
...
 }
 
-pub fn do_snapshot(snap: &RocksDbSnapshot,
+const RETRY_CNT: u32 = 1024;
+/// Name prefix for the self-generated snapshot file.
+pub const SNAP_GEN_PREFIX: &'static str = "gen";
+/// Name prefix for the received snapshot file.
+pub const SNAP_REV_PREFIX: &'static str = "rev";
+
+/// A structure represent the snapshot file.
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: represents

File: src/raftstore/store/peer_storage.rs [hunk 17]
```diff
@@ -503,9 +531,213 @@ impl PeerStorage {
...
+        let mut digest = Digest::new(crc32::IEEE);
+        let len = try!(reader.metadata()).len();
+        if len < 4 {
+            return Err(io::Error::new(ErrorKind::InvalidInput, format!("file length {} < 4", len)));
+        }
+        let to_read = len as usize - 4;
+        let mut total_read = 0;
+        let mut buffer = vec![0; 4098];
+        loop {
+            let readed = try!(reader.read(&mut buffer));
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: readed is not a correct word.

File: src/raftstore/store/peer_storage.rs [hunk 18]
```diff
@@ -503,9 +531,213 @@ impl PeerStorage {
...
+            snap_size += key.len();
+            snap_size += value.len();
+            snap_key_cnt += 1;
+
+            try!(f.encode_compact_bytes(key));
+            try!(f.encode_compact_bytes(value));
+            Ok(true)
+        }));
+    }
+    // use an empty byte array to indicate that kvpair reach an end.
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: reaches the end.

File: src/server/snap.rs [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+/// It will first send the normal raft snapshot message and then send the snapshot file.
+fn send_snap(snap_dir: PathBuf, addr: SocketAddr, data: ConnData) -> Result<()> {
+    assert!(data.is_snapshot());
+    let timer = Instant::now();
+    let snap = data.msg.get_raft().get_message().get_snapshot();
+    let mut snap_file = box_try!(SnapFile::from_snap(snap_dir.as_path(), &snap, SNAP_GEN_PREFIX));
+    if !snap_file.exists() {
+        return Err(box_err!("missing snap file: {:?}", snap_file.path()));
+    }
+    // snapshot file has been validated when created, so no need validate again.
```
Comment (siddontang) [inline src/server/snap.rs]: no need validating or no need to

File: tests/raftstore/node.rs [hunk 20]
```diff
@@ -16,35 +16,64 @@
...
+                Some(p) => SnapFile::from_snap(p.path(), snap, SNAP_GEN_PREFIX).unwrap(),
+                None => return Err(box_err!("missing temp dir for store {}", from_store)),
+            };
+            let mut dst_file = match self.snap_paths.get(&to_store) {
+                Some(p) => SnapFile::from_snap(p.path(), snap, SNAP_REV_PREFIX).unwrap(),
+                None => return Err(box_err!("missing temp dir for store {}", to_store)),
+            };
+            let mut reader = File::open(source_file.path()).unwrap();
+
+            dst_file.encode_u64(msg.compute_size() as u64).unwrap();
```
Comment (siddontang) [inline tests/raftstore/node.rs]: can we not use fs::copy directly?

Comment (BusyJay) [inline tests/raftstore/node.rs]: No, the snapshot file to be sent and the file finally saved to disk are not identical.

File: tests/raftstore/test_snap.rs [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    must_get_none(&engine_2, &format!("{:01024}", 0).into_bytes());
+    // add peer (2,2) to region 1.
+    pd_client.must_add_peer(r1, new_peer(2, 2));
+
+    let (key, value) = (b"k2", b"v2");
+    cluster.must_put(key, value);
+    assert_eq!(cluster.get(key), Some(value.to_vec()));
+    must_get_equal(&engine_2, key, value);
+
+    // now peer 2 must have v1 and v2;
```
Comment (siddontang) [inline tests/raftstore/test_snap.rs]: seem mismatched comment.

File: src/raftstore/store/peer_storage.rs [hunk 22]
```diff
@@ -503,9 +531,213 @@ impl PeerStorage {
     }
 }
 
-pub fn do_snapshot(snap: &RocksDbSnapshot,
+const RETRY_CNT: u32 = 1024;
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: no used const?

**Comment by siddontang** [comment]
Rest LGTM

**Comment by BusyJay** [comment]
@ngaut @disksing PTAL

File: src/server/snap.rs [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+    let timer = Instant::now();
+    let snap = data.msg.get_raft().get_message().get_snapshot();
+    let mut snap_file = box_try!(SnapFile::from_snap(snap_dir.as_path(), &snap, SNAP_GEN_PREFIX));
+    if !snap_file.exists() {
+        return Err(box_err!("missing snap file: {:?}", snap_file.path()));
+    }
+    // snapshot file has been validated when created, so no need to validate again.
+
+    let mut f = try!(File::open(snap_file.path()));
+    snap_file.delete_when_drop();
```
Comment (c4pt0r) [inline src/server/snap.rs]: I don't know if it's the common rule of rust-lang, but when I saw this method, I didn't realize it's a setter function, I thought it's a getter until I checked its implementation.
Do you guys have the same feeling?

File: src/server/snap.rs [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+    let mut conn = try!(TcpStream::connect(&addr));
+    try!(conn.set_nodelay(true));
```
Comment (c4pt0r) [inline src/server/snap.rs]: See, it's obvious a setter.

File: src/server/snap.rs (hunk 23 above)
Comment (BusyJay) [inline src/server/snap.rs]: Actually it's not a setter. It's just a function which command the snap_file to delete itself when it's dropped.

**Comment by ngaut** [comment]
LGTM

File: src/server/snap.rs (hunk 23 above)
Comment (c4pt0r) [inline src/server/snap.rs]: I am OK with that. It isn't a setter indeed

**Pull request merged**